xarray/core/combine.py

```diff
@@ -606,6 +606,15 @@ def vars_as_keys(ds):
     return tuple(sorted(ds))
 
 
+def groupby_defaultdict(iter, key=lambda x: x):
```

Comment:
```suggestion
T = TypeVar("T")
K = TypeVar("K", bound=Hashable)

def groupby_defaultdict(iter: list[T], key: Callable[[T], K] = lambda x: x) -> Iterator[tuple[K, Iterator[T]]]:
```
Could you add some type hints here?
Not sure if my suggestion is correct.

Reply:
Thanks @headtr1ck, but I have no idea what those `TypeVar` declarations are doing. The only thing I know is, that `iter` is a list of `xr.Dataset` and `key` is a function which consumes and returns `xr.Dataset`.

Reply:
Ah sorry,the key-function returns `tuple(sorted(xr.Datatset))`

Reply:
That's a tuple of strings, AFAICT.

Reply:
Ok, then it makes sense. The return value has to be hashable.

The TypeVars can be anything given the bounds (so K has to be hashable) but it has to be consistent in a function call.

E.g. the key argument needs to be the same type as the list contents.

Reply:
So, your suggestion will work then? Let's try.

Reply:
Oh, you need some imports from typing